### PR TITLE
avoid thunk allocation in aroundReceive, #20748

### PR DESCRIPTION
**Pull request by patriknw**
- had to cache the function in ActorCell due to binary compatibility
  limitation

### CODE REVIEW

**Comment by johanandren** [comment]
LGTM

**Comment by ktoso** [comment]
Refs https://github.com/akka/akka/issues/20748

File: akka-actor/src/main/scala/akka/actor/Actor.scala [hunk 1]
```diff
@@ -442,6 +442,12 @@ trait Actor {
           "You have to use one of the 'actorOf' factory methods to create a new actor. See the documentation.")
     val c = contextStack.head
     ActorCell.contextStack.set(null :: contextStack)
+    // cache function to `unhandled` to avoid allocations in aroundReceive,
+    // can't use a val in Actor due to binary compatibility restriction
+    c match {
+      case cell: ActorCell ⇒ cell.unhandledFun = unhandled
+      case _               ⇒
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/Actor.scala]: I saw the discussion on gitter about why we don't care - that's fine :)
Please add a small comment here why we don't though

File: akka-actor/src/main/scala/akka/actor/ActorCell.scala [hunk 2]
```diff
@@ -412,6 +412,13 @@ private[akka] class ActorCell(
     unstashed
   }
 
+  /**
+   * INTERNAL API
+   * Cached function to unhandled to avoid allocations in aroundReceive,
+   * can't use a val in Actor due to binary compatibility restriction
+   */
+  private[akka] var unhandledFun: Any ⇒ Unit = null
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/ActorCell.scala]: I should probably clear this field in `clearActorCellFields`

File: akka-actor/src/main/scala/akka/actor/Actor.scala (hunk 1 above)
Comment (drewhk) [inline akka-actor/src/main/scala/akka/actor/Actor.scala]: maybe with an extra explanation when that can actually happen

File: akka-actor/src/main/scala/akka/actor/ActorCell.scala (hunk 2 above)
Comment (drewhk) [inline akka-actor/src/main/scala/akka/actor/ActorCell.scala]: yes, I think so, too

**Comment by drewhk** [comment]
LGTM

**Comment by ktoso** [comment]
LGTM after minor nitpicks addressed 👍

**Comment by patriknw** [comment]
That failure is scary, `SupervisorHierarchySpec`. I got it failing once on my local also, but only once. Hmm, have I misunderstood anything about lifting functions like this?

**Comment by patriknw** [comment]
now, it failed one more time on local

**Comment by drewhk** [comment]
Unsafe publication? I am not sure at the location where you set the value it is guaranteed that it will become visible. Just a tip though.

**Comment by patriknw** [comment]
Good thought, but shouldn't it be the same due to the fallback when null?

**Comment by drewhk** [comment]
Dunno, it I didn't look at the error. Might be partially constructed object being visible, but it might be more mundane.

**Comment by patriknw** [comment]
New attempt

**Comment by patriknw** [comment]
this time a long GC in CoderSpec

**Comment by patriknw** [comment]
PLS BUILD

**Comment by patriknw** [comment]
I will check if the allocation is removed in Scala 2.12-M5, when that is released.

**Comment by patriknw** [comment]
closing until we have tested with Scala 2.12-M5

**Pull request closed without merging**